plugins/ltac/pltac.ml

```diff
@@ -66,4 +66,6 @@ let () =
   register_grammar wit_ltac (tactic);
   register_grammar wit_clause_dft_concl (clause_dft_concl);
   register_grammar wit_destruction_arg (destruction_arg);
+  register_grammar Genredexpr.wit_red_expr (Pvernac.Vernac_.red_expr);
```

Comment:
`wit_red_expr` should also be moved up, since it's not used outside of Ltac.

Reply:
But then, most of `stdarg.ml` could go to Ltac too (only `Auto.conclPattern` seems to need a `wit`), right?

Reply:
My personal take on that is that the whole generic argument mechanism should be moved into Ltac, except for VERNAC-only arguments (i.e. those which do not use their glob / typed parameters). Currently we abuse a bit the system, which allows more code sharing between Vernac and Ltac, but these two kinds of parameters are really distinct, and most of the code is Ltac-specific. Vernac arguments only use parsing and printing IIRC.

For the time being it doesn't matter, but since you're touching this part of the code I proposed to `wit_red_expr` as well.

Reply:
So, are we considering `red_expr` and `binders` as liable to be used in vernac extensions? It seems to me yes.

Reply:
At the risk of repeating myself, red_expr is Ltac syntax.

Reply:
Yes, we are repeating ourselves. Is `Eval compute in foo` ltac syntax or gallina not? It is not about tactics at all, so why should it be considered ltac syntax?

Is it that you intend eventually to provide another syntax for `Eval red_expr in foo` and would like it not to be told to be a gallina syntax?

I believe that `Eval compute in foo`, or at least something similar, should be available even when `Ltac` is not required. What would you recommend otherwise to someone which wants to use `red_expr` in a new vernac commands? That they have to require ltac mandatorily?

Reply:
> why should it be considered ltac syntax?

You can pass variables to some reduction functions, and those are interpreted according to the Ltac "semantics". The very fact we have non-trivial glob / typed parameters is a red flag. Only a strict subset of the redexpr syntax qualifies as vernacular.

Reply:
I agree that the architecture interpreting red_expr supports variables, but that the same for constr, and indeed, pretyping also "supports" ltac variables.

Can't we consider that the support of variables in pretyping or in red_expr is a support for "meta-language variables" which has nothing to do with Ltac per se but more generally to meta-programming.

Conversely what are you calling the ``Ltac "semantics"'', which would make that these meta-language variables have to be special to only Ltac?

Reply:
> Can't we consider that the support of variables in pretyping or in red_expr is a support for "meta-language variables" which has nothing to do with Ltac per se but more generally to meta-programming?

No, for instance Ltac2 does not rely on this mechanism. Just reading the code of `Pretyping.pretype_id` should fill you with dread.

Reply:
Afaics, interpretation of Ltac2 variables in terms dynamically add a syntax `$x` which is interpreted as an `ltac2:(x)` (i.e. a `GHole`) itself being interpreted in pretyping by calling an Ltac2-provided interpretation function which interprets the term `x` in a typing environment encoded into a special variable name `@@ltac2_env@@`. So, basically, this is already what `pretype_id` is already providing, without the `@@ltac2_env@@` encoding. 

The difference is that Ltac uses the same name space for term variables and Ltac variables while Ltac2 uses distinct name spaces. So it cannot reuse `pretype_id` and needs a distinct node in `glob_constr`.

I'm not convinced that the convention that term variables and LtacX variables have to leave in different name spaces is important. It is a question on the convention on what statically binds what and I would tend to think that the Ltac interpretation could also choose to rely on two name spaces (and use a `GHole` encoding too), with interning deciding when a variable is a term variable (in the current constr env) or an Ltac variable (in the ist) and that would work the same (up to `tacinterp` probably bypassing sometimes the decision of `tacintern`). The question is more about how to correctly propagate at interning time the list of known variable names that an Ltac variable is supposed to find in the constr context at runtime.

Independently, a new explicit constructor in `glob_constr` for Ltac2 variables could be introduced rather than using a `GHole` encoding.

Reply:
> So, basically, this is already what pretype_id is already providing, without the @@ltac2_env@@ encoding.

No. When you see this node, you *statically* know that this is a quotation. When you see an Ltac variable, you have to decide *at runtime* what this might be, with interesting consequences sometimes.

> I'm not convinced that the convention that term variables and LtacX variables have to leave in different name spaces is important.

I know you fancy it, but this is literally the bane of Ltac1. This is the feature that ultimately prevents adding a static type system on Ltac1, because you must defer type-checking to runtime to check whether a variable must come from the dynamic context or be a well-typed bound metavariable.

Reply:
> you have to decide `at runtime`

In fine, not more than in Ltac2 which dynamically type-checks metavariables in a "random" context. In particular, I claim that there is a model of Ltac in an extended Ltac2 where all variables are metavariables, that is that `Ltac f x := intros y; apply (fun z => x + y + z)` is interpreted as `Ltac2 f x := intros y; apply (fun z => $x + $y + $z)`, where the "extended" is that any binder in terms and introduction tactic implicitly generates an associated metavariable of union type name+constr. Without having explored more details than I did in [#11721](https://github.com/coq/coq/pull/11721#issuecomment-595422772) though, I claim that such interpretation can be equipped with a type system which correctly tracks the dependencies (and an example such as [below](https://github.com/coq/coq/pull/13396#issuecomment-851698962) would be rejected with error `unbound variable y` in `clear y`).

Reply:
If your objective is to move `globEnv` to `Ltac` and have `pretype_id` only look in the constr environment, I believe this is possible by doing so:
- at interning time of an Ltac expression, turn any variable to a `GHole` ltac quotation (which in turn will implement  the current `pretype_id` at the Ltac level)
- add the ability to apply an effect at each `push_rel` in `pretype`, this effect being registered from Ltac (this would give the equivalent of current `GlobEnv.push_rel`).

Reply:
To come back on `wit_red_expr`, I don't know what to conclude. If a new vernac refers to a `red_expr`, the Ltac interpretation part will be supplied with an empty environment (in `Tacinterp.interp_redexpr`) and this will work as if Ltac were not involved. So, whatever we may think of the semantics of Ltac, this will not interfere with a potential use in vernac extensions.

Alternatively, if we want to make the interpretation of `red_expr` in vernac independent of the Ltac variables, I don't know how to do it without duplicating the code.